File: packages/schemas-python/src/prism_schemas/verdict.py
"""Sentinel verdict schema."""

from __future__ import annotations

import hashlib
import json
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


ChallengeType = Literal[
    "temporal",
    "source_quality",
    "relevance",
    "calibration",
    "logic",
    "market_structure",
    "risk",
]
ChallengeSeverity = Literal["minor", "material", "blocking"]
ChallengeResolutionStatus = Literal["open", "answered", "resolved", "conceded", "superseded"]
ResolutionStopReason = Literal[
    "single_shot",
    "confidence_reached",
    "max_rounds",
    "unresolved_blockers",
    "insufficient_evidence",
]


class AdversarialChallenge(BaseModel):
    """A structured issue raised by the sentinel against a trace."""

    id: str
    type: ChallengeType
    severity: ChallengeSeverity
    question: str
    required_resolution: str
    blocking_pass: bool = False
    claim_ref: str | None = None
    resolution_status: ChallengeResolutionStatus = "open"


class EvidenceToolReceipt(BaseModel):
    """Structured provenance for evidence used to resolve an issue."""

    provider: str
    tool_name: str | None = None
    source_title: str
    source_url: str
    source_published_at: str | None = None
    retrieved_at: str | None = None
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    adequacy_checks: list[str] = Field(default_factory=list)


class ChallengeResolution(BaseModel):
    """Trader or tool response that attempts to resolve one challenge."""

    challenge_id: str
    status: ChallengeResolutionStatus
    responder: Literal["trader", "sentinel", "evidence_tool", "system"]
    response: str
    created_at: datetime
    tool_receipt: EvidenceToolReceipt | None = None


class ResolutionRound(BaseModel):
    """One bounded adversarial dialogue round."""

    round_index: int = Field(ge=0)
    opened_challenge_ids: list[str]
    resolved_challenge_ids: list[str]
    prompt: str
    response: str
    created_at: datetime


class AdversarialResolutionMetadata(BaseModel):
    """Summary of the adversarial resolution loop state."""

    confidence: float = Field(ge=0.0, le=1.0)
    stop_reason: ResolutionStopReason
    unresolved_blocking_count: int = Field(ge=0)
    unresolved_material_count: int = Field(ge=0)
    max_rounds: int = Field(ge=0)


class SentinelVerdict(BaseModel):
    """Adversarial validation verdict from the sentinel."""

    request_hash: str
    trace_id: str
    sentinel_agent_id: int

    evidence_challenges: list[str]
    thesis_challenges: list[str]
    calibration_critique: str

    verdict_score: int = Field(ge=0, le=100)
    verdict_label: Literal["REJECT", "WARN", "PASS", "ENDORSE"]

    dialogue_messages: list[dict[str, Any]]

    structured_challenges: list[AdversarialChallenge] = Field(default_factory=list)
    challenge_resolutions: list[ChallengeResolution] = Field(default_factory=list)
    resolution_rounds: list[ResolutionRound] = Field(default_factory=list)
    resolution_metadata: AdversarialResolutionMetadata | None = None

    model_family: Literal["anthropic-claude", "openai-gpt"]
    model_name: str
    created_at: datetime

    requester_address: str | None = None

    def content_hash(self) -> bytes:
        """Deterministic hash for on-chain anchoring.

        `exclude_defaults=True` preserves the canonical bytes of older pinned
        verdict JSON that did not contain newer optional fields. New verdicts
        that set non-default issue-ledger fields still include them in the hash.
        """
        canonical = json.dumps(
            self.model_dump(mode="json", exclude_defaults=True),
            sort_keys=True,
        )
        return hashlib.sha256(canonical.encode()).digest()

File: packages/schemas-typescript/src/verdict.ts
/**
 * Sentinel verdict schema — Zod v4 mirror of Python Pydantic model.
 */

import { z } from "zod/v4";

/** A dialogue message from the sentinel's adversarial review. */
export const DialogueMessageSchema = z.object({
  role: z.string(),
  content: z.string(),
});

export type DialogueMessage = z.infer<typeof DialogueMessageSchema>;

export const ChallengeTypeSchema = z.enum([
  "temporal",
  "source_quality",
  "relevance",
  "calibration",
  "logic",
  "market_structure",
  "risk",
]);

export const ChallengeSeveritySchema = z.enum(["minor", "material", "blocking"]);

export const ChallengeResolutionStatusSchema = z.enum([
  "open",
  "answered",
  "resolved",
  "conceded",
  "superseded",
]);

export const ResolutionStopReasonSchema = z.enum([
  "single_shot",
  "confidence_reached",
  "max_rounds",
  "unresolved_blockers",
  "insufficient_evidence",
]);

export const AdversarialChallengeSchema = z.object({
  id: z.string(),
  type: ChallengeTypeSchema,
  severity: ChallengeSeveritySchema,
  question: z.string(),
  required_resolution: z.string(),
  blocking_pass: z.boolean().default(false),
  claim_ref: z.string().nullable().optional(),
  resolution_status: ChallengeResolutionStatusSchema.default("open"),
});

export type AdversarialChallenge = z.infer<typeof AdversarialChallengeSchema>;

export const EvidenceToolReceiptSchema = z.object({
  provider: z.string(),
  tool_name: z.string().nullable().optional(),
  source_title: z.string(),
  source_url: z.string(),
  source_published_at: z.string().nullable().optional(),
  retrieved_at: z.string().nullable().optional(),
  confidence: z.number().min(0).max(1).default(0),
  adequacy_checks: z.array(z.string()).default([]),
  extractor_provider: z.string().nullable().optional(),
  extractor_tool_name: z.string().nullable().optional(),
  source_content_hash: z.string().nullable().optional(),
  source_excerpt: z.string().nullable().optional(),
  extracted_at: z.string().nullable().optional(),
  extraction_checks: z.array(z.string()).default([]),
});

export type EvidenceToolReceipt = z.infer<typeof EvidenceToolReceiptSchema>;

export const ChallengeResolutionSchema = z.object({
  challenge_id: z.string(),
  status: ChallengeResolutionStatusSchema,
  responder: z.enum(["trader", "sentinel", "evidence_tool", "system"]),
  response: z.string(),
  created_at: z.string(),
  tool_receipt: EvidenceToolReceiptSchema.nullable().optional(),
});

export type ChallengeResolution = z.infer<typeof ChallengeResolutionSchema>;

export const ResolutionRoundSchema = z.object({
  round_index: z.number().int().min(0),
  opened_challenge_ids: z.array(z.string()),
  resolved_challenge_ids: z.array(z.string()),
  prompt: z.string(),
  response: z.string(),
  created_at: z.string(),
});

export type ResolutionRound = z.infer<typeof ResolutionRoundSchema>;

export const AdversarialResolutionMetadataSchema = z.object({
  confidence: z.number().min(0).max(1),
  stop_reason: ResolutionStopReasonSchema,
  unresolved_blocking_count: z.number().int().min(0),
  unresolved_material_count: z.number().int().min(0),
  max_rounds: z.number().int().min(0),
});

export type AdversarialResolutionMetadata = z.infer<typeof AdversarialResolutionMetadataSchema>;

/** Adversarial validation verdict from the sentinel. */
export const SentinelVerdictSchema = z.object({
  request_hash: z.string(),
  trace_id: z.string(),
  sentinel_agent_id: z.number(),

  evidence_challenges: z.array(z.string()),
  thesis_challenges: z.array(z.string()),
  calibration_critique: z.string(),

  verdict_score: z.number().int().min(0).max(100),
  verdict_label: z.enum(["REJECT", "WARN", "PASS", "ENDORSE"]),

  dialogue_messages: z.array(DialogueMessageSchema),

  structured_challenges: z.array(AdversarialChallengeSchema).default([]),
  challenge_resolutions: z.array(ChallengeResolutionSchema).default([]),
  resolution_rounds: z.array(ResolutionRoundSchema).default([]),
  resolution_metadata: AdversarialResolutionMetadataSchema.nullable().optional(),

  model_family: z.enum(["anthropic-claude", "openai-gpt"]),
  model_name: z.string(),
  created_at: z.string(),
});

export type SentinelVerdict = z.infer<typeof SentinelVerdictSchema>;
